Sources/Sentry/SentryANRTracker.m

```diff
@@ -0,0 +1,110 @@
+#import "SentryANRTracker.h"
+#import "SentryCrashAdapter.h"
+#import "SentryDispatchQueueWrapper.h"
+#import "SentryLog.h"
+#import "SentryThreadWrapper.h"
+#import <Foundation/Foundation.h>
+
+@interface
+SentryANRTracker ()
+
+@property (weak, nonatomic) id<SentryANRTrackerDelegate> delegate;
+@property (nonatomic, assign) NSTimeInterval timeoutInterval;
+@property (nonatomic, strong) id<SentryCurrentDateProvider> currentDate;
+@property (nonatomic, strong) SentryCrashAdapter *crashAdapter;
+@property (nonatomic, strong) SentryDispatchQueueWrapper *dispatchQueueWrapper;
+@property (nonatomic, strong) SentryThreadWrapper *threadWrapper;
+
+@property (weak, nonatomic) NSThread *thread;
+
+@end
+
+@implementation SentryANRTracker
+
+- (instancetype)initWithDelegate:(id<SentryANRTrackerDelegate>)delegate
+           timeoutIntervalMillis:(NSUInteger)timeoutIntervalMillis
+             currentDateProvider:(id<SentryCurrentDateProvider>)currentDateProvider
+                    crashAdapter:(SentryCrashAdapter *)crashAdapter
+            dispatchQueueWrapper:(SentryDispatchQueueWrapper *)dispatchQueueWrapper
+                   threadWrapper:(SentryThreadWrapper *)threadWrapper
+{
+    if (self = [super init]) {
+        self.delegate = delegate;
+        self.timeoutInterval = (double)timeoutIntervalMillis / 1000;
+        self.currentDate = currentDateProvider;
+        self.crashAdapter = crashAdapter;
+        self.dispatchQueueWrapper = dispatchQueueWrapper;
+        self.threadWrapper = threadWrapper;
+    }
+    return self;
+}
+
+- (void)start
+{
+    [NSThread detachNewThreadSelector:@selector(detectANRs) toTarget:self withObject:nil];
+}
+
+- (void)detectANRs
+{
+    NSThread.currentThread.name = @"io.sentry.anr-tracker";
+
+    self.thread = NSThread.currentThread;
+
+    BOOL wasPreviousANR = NO;
+
+    while (![self.thread isCancelled]) {
+
+        NSDate *blockDeadline =
+            [[self.currentDate date] dateByAddingTimeInterval:self.timeoutInterval];
+
+        __block BOOL blockExecutedOnMainThread = NO;
+        [self.dispatchQueueWrapper dispatchOnMainQueue:^{ blockExecutedOnMainThread = YES; }];
+
+        [self.threadWrapper sleepForTimeInterval:self.timeoutInterval];
+
+        if (blockExecutedOnMainThread) {
+            if (wasPreviousANR) {
+                [SentryLog logWithMessage:@"ANR stopped." andLevel:kSentryLevelDebug];
```

Comment:
When we detect an ANR (and detect it stopped) what about we log with Warning (or perhaps Info) level?

```suggestion
                [SentryLog logWithMessage:@"ANR stopped." andLevel:kSentryLevelWarning];
```

I'm afraid otherwise it'll be "all or nothing" in the diagnostic logging: Debug you get everything (and a lot of noise), anything else won't give you enough info to understand what's going on.